flojoy/flojoy_node_venv.py

```diff
@@ -0,0 +1,142 @@
+import venv
+from functools import  wraps
+
+
+__all__ = ["run_in_venv"]
+
+class MultiprocessingExecutableContextManager:
+    """Temporarily change the executable used by multiprocessing."""
+    def __init__(self, executable_path):
+        self.original_executable_path = sys.executable
+        self.executable_path = executable_path
+
+    def __enter__(self):
+        self.original_start_method = multiprocessing.get_start_method()
+        multiprocessing.set_executable(self.executable_path)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        multiprocessing.set_executable(self.original_executable_path)
+
+class SwapSysPath:
+    """Temporarily swap the sys.path of the child process with the sys.path of the parent process."""
+    def __init__(self, venv_executable):
+        self.new_path = _get_venv_syspath(venv_executable)
+        self.old_path = None
+
+    def __enter__(self):
+        self.old_path = sys.path
+        sys.path = self.new_path
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.path = self.old_path
+
+
+def _install_pip_dependencies(
+        venv_executable: os.PathLike,
+        pip_dependencies: tuple[str],
+        verbose: bool = False
+    ):
+    """Install pip dependencies into the virtual environment."""
+    command = [venv_executable, "-m", "pip", "install"]
+    if(not verbose):
+        command += ["-q"]  
+    command += list(pip_dependencies)
+    subprocess.check_call(command)
+
+def _get_venv_syspath(venv_executable: os.PathLike) -> list[str]:
+    """Get the sys.path of the virtual environment."""
+    command = [venv_executable, "-c", "import sys\nprint(sys.path)"]
+    cmd_output = subprocess.run(command, check=True, capture_output=True, text=True)
+    return eval(cmd_output.stdout)
+
+class PickleableFunctionWithPipeIO:
+    """A wrapper for a function that can be pickled and executed in a child process."""
+    def __init__(self, func, child_conn, venv_executable):
+        self._func_serialized = cloudpickle.dumps(func)
+        self._child_conn = child_conn
+        self._venv_executable = venv_executable
+
+    def __call__(self, *args, **kwargs):
+        fn = cloudpickle.loads(self._func_serialized)
+        args = [cloudpickle.loads(arg) for arg in args]
+        kwargs = {key: cloudpickle.loads(value) for key, value in kwargs.items()}
+        with SwapSysPath(venv_executable=self._venv_executable):
+            result = fn(*args, **kwargs)
+        self._child_conn.send(result)
+
+def run_in_venv(pip_dependencies=[]):
+    # Pre-pend flojoy and cloudpickle as mandatory pip dependencies
+    pip_dependencies = sorted(["flojoy", "cloudpickle"] + pip_dependencies)
+    # Get the system's temporary directory path
+    temp_dirpath = tempfile.gettempdir()
+    # Generate a path-safe hash of the pip dependencies
+    # this prevents the duplication of virtual environments
+    pip_dependencies_hash = hashlib.sha256("".join(pip_dependencies).encode()).hexdigest()
```

Comment:
I find this solution neat, but I’d like to throw an idea here: what if the users could pass a venv name as an optional identifier? 
Installing venvs on the fly will hurt performance, and if the users are led to fill in only the strictly necessary requirements for each node, what we’ll have at the end of the day is one venv per node per pipeline. On the other hand, I can think of many cases where a group of nodes within the same execution pipeline would run just fine within the same venv, even if individually these nodes have slightly different minimal requirements on their own.

Of course, the users could do this in the current implementation by passing the same pip_dependencies list to all nodes, but somehow this feels less intuitive to me than choosing the same venv name.

The venv_name could be an optional arg and we could then use the hash as uuid if the user does not choose one.

Reply:
I can 100% see the benefit of shared venvs, as we don't want to re-download and re-build packages if already present in other environments.
However I think it poses challenges to manage these shared environments:

(1) It breaks the isolation we initially wanted to bring with venvs
(2) Reproducibility is harder to guarantee and we have less visibility in "what is in this environment". 
    We need to go to different places to find that out, then after a modification is applied, we need to test
    for all the nodes that use the shared venv
(3) An error in a virtual environment can now propagate to multiple nodes

The good news is that `pip` has a system-wide cache, which means if a package version is present in another existing environment and if it is specified in a new environment, then `pip` will copy it from the cache and unpack it in the new environment, and I hope this gets is 75% the way to where we wanted to go initially.

CAVEAT: The current implementation saves the virtual environments in `/tmp`, and so they are not persisted. 
WDYT about moving it to something persistent such as `~/.flojoy/.cache/` for example?

SUGGESTION: When receiving a new venv dependency list, WDYT about scanning for all existing venvs to see if there exists another venv whose dependencies are a superset of the new venv about to be created?

Reply:
> WDYT about moving it to something persistent such as ~/.flojoy/.cache/ for example?

this is definitely an improvement, it may allow the Flojoy app to manage the existing environments, re-use previously created ones, delete them accordindly, etc...

Reply:
I applied the modification, PTAL @luiztauffer

Reply:
This PR was merged, please follow up with in slack/pr/issue if appropriate